Add config inclusion feature

sebastianfeldmann opened the pull request:
Hi @heiglandreas 

Could you have a look at this, and let me know what you think?

This adds the possibility to include `CaptainHook` configurations in `CaptainHook` configuration files.
Basically it would look like this

```javascript
{
  "config": {
    "includes": [
      "vendor/my-company/hook-package/valid-include.json"
    ]
  },
  ...
  "pre-commit": {
    "enabled": true,
    "actions": [
      {
        "action": "phpunit --configuration=build/phpunit-hook.xml",
        "options": []
      }
    ]
  },
  ...
}
```

If the included configuration would look like this
```javascript
{
  "pre-commit": {
    "enabled": true,
    "actions": [
      {
        "action": "phpcs --standard=psr2 src",
        "options": []
      }
    ]
  }
}
```

and you run `CaptainHook` with the first configuration both `pre-commit` actions would be executed.

This way you could create  a "company base packages" that requires `CaptainHook` and maybe some plugins. Also the package could provide some base configurations that can be included in all project `captainhook.json` files.

This way if you want to add some validation to all projects the only thing you have to do is to update the package and run `composer update` on all projects.

I have not yet tested if a package can actually `require` a plugin, but I would hope that it's possible.

heiglandreas commented:
Looks awesome! I'll dig into that tomorrow morning...

heiglandreas commented:
The idea is great! But I do see some issues with the current implementation (that can be worked around IMO)

* First, as far as I see it, the current `captainhook.json` is not modified. Which means that the inclusion of the file is done on each request of the File, which is on each and every invocation of a git hook. That can be rather time-consuming. Wouldn't it be easier/faster to create a `captainhook.lock` file (or whatever it is called) that contains the merged config which is then used instead of the `captainhook.json`? That `.lock`-file is created on every `install` or `update` as those are the only possibilities that the content can change. Alternatively when the lock` file is not available it will be created on the fly, which would mean that the `install` or `update` only needs to remove that `.lock`-file. Benefit would be, that it would be possible to load JSON from websites as well (with all the possible attack-vectors that suddenly opens up)
* Secondly I'm not sure how you would use that. Would you always `require --dev` one company-package that then includes a project specific `captainhook.json`? Or would you set up The Captain individually with each project and then include a companywide `captainhook.json`-file? The first would require dependencies to install composer plugins while the later would list all the dependencies the project uses right in the right place. When using the first approach that would mean that the company-dependency would need to require The Captain and possible plugins as dependencies and not dev-dependencies as they would not be loaded otherwise. In the later case one could require a company-wide `captainhook.json` right from a webserver (given we allow such a thing – see first point).

I think those are my main concerns.

sebastianfeldmann commented:
Thanks for taking the time to look at this
So let´s address your concerns.

## CaptainHook lock file

I terms of the `captainhook.lock` file idea. I'm not sure I want to go there yet. Handling `lock` files is quite a bit of work. Creating hashes to ensure the validity of those files writing them to disk...
Composer is prove that it could be done reliably, but I want to avoid the complexity for now.
I benchmarked the config loading process by loading  a configuration with 2 included configuration files containing roughly 1k actions for different hooks each.

    Loaded 3 configs in 0.0069s

For now I don't think loading the configuration every time is an issue in terms of performance we have to worry about. I would argue that in realistic scenarios one would include 2 or 3 configuration files with maybe 3 to 5 actions. Loading such a realistic configuration with 3 includes and an overall amount of 23 actions looked like this.

    Loaded 4 configs in 0.0024s

 ## General purpose

Every project still keeps its own `captainhook.json` to specify project specific settings. But as @aboks pointed out on twitter, if you have a lot of projects and want the same configuration for all of them, currently you have to manage it on copy&paste basis or write custom actions and plugins.

The idea is that you could create a basic `composer` package like this.

```
my-git-hooks/
  |-- composer.json
  |-- msg-hooks.json
  |-- crazy-workflow-hooks.json
```

Within its `composer.json` you would require (NOT `--dev`) all libs and plugins you need for the configured hooks in your configuration files (`msg-hooks.json`, `crazy-workflow-hooks.json`) as well as `captainhook/captainhook` or `captainhook/plugin-composer`.

To use your basic configuration you would have to require your basic package as `--dev` dependency.

    composer require --dev my-company/my-git-hooks

Then create or edit the project `captainhook.json` and put something in like the following.

```javascript
{
  "config": {
    "includes": [
      "vendor/my-company/my-git-hooks/msg-hook.json",
      "vendor/my-company/my-git-hooks/crazy-workflow-hooks.json"
    ]
  },
  ...
}
```

To update the base configuration you create a new version for your `my-git-hooks` package and run `composer update` in your projects.
 
### HTTP handling

Opening this up to URLs would be problematic on many fronts.

1. The performance issue you mentioned before would get a lot more serious if we would do http instead of local files.
2. It would not be possible to require dependencies via http so you would have to figure that out in your local `composer.json` and manage all the dependency dependencies there. Not ideal to say at least.

So for now I would keep it as local file inclusion only feature.

## To do

In this prove of concept implementation it currently works recursively, so you could include configurations in your included configurations. I'm not sure this is a good idea. No, I know this is not a good idea, I just had no time to make it 1 level of inclusion only. But that should be done before we release it.

aboks commented on src/Config/Factory.php:
```diff
@@ -98,4 +100,54 @@ protected function configureHook(Config\Hook $config, array $json) : void
+    }
+
+    /**
+     * Append actions for a given hook from all included configurations to the current configuration
```
I think that "from a single included configuration" rather than "from all included configurations" would be a more accurate comment here.

aboks commented:
Thanks for your work on this. I think @sebastianfeldmann already described my envisoned usecase better than I could have myself, and the current implementation seems perfectly suitable for that. 

Personally I don't mind taking a minor performance hit when using included configs. As long as the case without included configs maintains almost its original performance, people can choose for themselves if they find the flexibility of included configs worth the slight performance hit.

As for recursive inclusion of configs, I personally probably won't use that, so feel free to limit it to 1 level. On the other hand, there probably isn't much harm in recursive inclusion. If people want to shoot themselves in the foot, they will always find a way to do so. If someone really wants to use 4 levels of configuration (and accepts the time necessary to load these), why prevent that?

heiglandreas commented:
Thanks for correcting me. The problem with the missing code when including remote URLs was something I didn't recognize in the first place. And after the clarification I'm absolutely with the current implementation. The performance issue seems to be negligible so that's OK. And should it become a problem it can then be addressed.

Only thing I ask myself is whether we can use composer to automatically include the files. Like creating a special package type that requires a special entry in their composer.json that can then be read by captain hook. But that is also an improvement that should not hold back the base implementation.

sebastianfeldmann commented:
> Only thing I ask myself is whether we can use composer to automatically include the files. Like creating a special package type that requires a special entry in their composer.json that can then be read by captain hook. But that is also an improvement that should not hold back the base 
implementation.

That would be nice

I added a config setting `maxIncludeLevel` that defaults to 1. So per default I try to keep everybody from shooting themselves in the foot, but if they insist they can do it.
You can set it like this

```javascript
{
  "config": {
    "maxIncludeLevel": 2,
    "includes": [
      "some-config.json"
    ]
    ...
}
```

From now on `CaptainHook` fails if an included configuration is not available and additionally I added the `@internal` annotation to all config related classes.
